python/taichi/lang/__init__.py

```diff
@@ -105,83 +139,57 @@ def init(arch=None,
         ti.get_runtime().set_default_fp(default_fp)
     if default_ip is not None:
         ti.get_runtime().set_default_ip(default_ip)
-    if print_preprocessed is not None:
-        ti.get_runtime().print_preprocessed = print_preprocessed
-
-    if debug is None:
-        debug = bool(int(os.environ.get('TI_DEBUG', '0')))
-    if debug:
-        ti.set_logging_level(ti.TRACE)
-    ti.cfg.debug = debug
-
-    unified_memory = os.environ.get('TI_USE_UNIFIED_MEMORY', '')
-    if unified_memory != '':
-        use_unified_memory = bool(int(unified_memory))
-        ti.cfg.use_unified_memory = use_unified_memory
-        if not use_unified_memory:
-            ti.trace(
-                'Unified memory disabled (env TI_USE_UNIFIED_MEMORY=0). This is experimental.'
-            )
-
-    for k, v in kwargs.items():
-        setattr(ti.cfg, k, v)
 
-    def bool_int(x):
-        return bool(int(x))
-
-    def environ_config(key, cast=bool_int):
-        name = 'TI_' + key.upper()
-        value = os.environ.get(name, '')
-        if len(value):
-            setattr(ti.cfg, key, cast(value))
-
-        # TI_ASYNC=   : not work
-        # TI_ASYNC=0  : False
-        # TI_ASYNC=1  : True
-
-    # does override
-    environ_config("print_ir")
-    environ_config("verbose")
-    environ_config("fast_math")
-    environ_config("async")
-    environ_config("print_benchmark_stat")
-    environ_config("device_memory_fraction", float)
-    environ_config("device_memory_GB", float)
-
-    # Q: Why not environ_config("gdb_trigger")?
-    # A: We don't have ti.cfg.gdb_trigger yet.
-    # Discussion: https://github.com/taichi-dev/taichi/pull/879
-    gdb_trigger = os.environ.get('TI_GDB_TRIGGER', '')
-    if len(gdb_trigger):
-        ti.set_gdb_trigger(bool(int(gdb_trigger)))
-
-    advanced_optimization = os.environ.get('TI_ADVANCED_OPTIMIZATION', '')
-    if len(advanced_optimization):
-        ti.core.toggle_advanced_optimization(bool(int(advanced_optimization)))
-
-    # Q: Why not environ_config("arch", ti.core.arch_from_name)?
-    # A: We need adaptive_arch_select for all.
-    env_arch = os.environ.get("TI_ARCH")
+    # configurations that are not in ti.cfg:
+    env.add('print_preprocessed', False)
+    env.add('log_level', 'info', str)
+    env.add('gdb_trigger', False)
+    env.add('excepthook', False)
+
+    # configurations in ti.cfg:
+    env.cfg_keys = []
+    env.add('debug')
+    env.add('print_ir')
+    env.add('verbose')
+    env.add('fast_math')
+    env.add('async')
+    env.add('use_unified_memory')
+    env.add('print_benchmark_stat')
+    env.add('advanced_optimization')
+    # TODO(yuanming-hu): Maybe these CUDA specific configs should be moved
+    # to somewhere like ti.cfg.cuda so that user don't get confused?
+    env.add('device_memory_fraction', None, float)
+    env.add('device_memory_GB', None, float)
+    for key in env.cfg_keys:
+        value = env[key]
+        if value is not None:
+            setattr(ti.cfg, key, value)
```

Comment:
> `if value is not None`

@k-ye Your concern is killed here.

Reply:
See? We **won't override the ti.cfg.x value** when `default=None`. @k-ye

Reply:
I'm not saying that this doesn't work, i was  **questioning the need for `default` arg**. IMO `_EnvironConfig` should just work for `ti.cfg`, which cleans up the design.

> This sounds too ad-hoc to me

Yes, but the existing `env.cfg_keys = []` approach is just equally ad-hoc. And especially now it has to process non `ti.cfg` keys like `print_preprocessed ` twice. The problem is that `_EnvironConfig ` is doing two things here: reading the env vars and applying **a subset of them** to `ti.cfg`. If we start mixing the env vars for `ti.cfg` and other subsystems like logging or GDB, this messy logic is inevitable. That's why the old `environ_config()` function just works for `ti.cfg`.

Reply:
So what should I treat with non-`ti.cfg` keys? Back to `init(print_processed=False)` again?